BFF Config Documentation (#136)
trisacrypto/trisa: docs/content/gds/admin/configuration/bff.en.md

| GDS_BFF_MODE          | string   | release               | Sets the Gin mode, one of debug, release, or test.                                                               |
| GDS_BFF_LOG_LEVEL     | string   | info                  | The verbosity of logging, one of trace, debug, info, warn, error, fatal, or panic.                               |
| GDS_BFF_CONSOLE_LOG   | bool     | false                 | If true will print human readable logs instead of JSON logs for machine consumption.                             |
| GDS_BFF_ALLOW_ORIGINS | []string | http://localhost:3000 | A list of allowed origins for the CORS middleware to accept.                                                     |
| GDS_BFF_REGISTER_URL  | string   |                       | The base URL to direct users to for registration signup (no trailing slash) - used in email templates.           |
| GDS_BFF_LOGIN_URL     | string   |                       | The base URL to direct users to for login (no trailing slash) - used in email templates.                         |

> | GDS_BFF_COOKIE_DOMAIN | string   |                       | The domain to set secure cookies for (particularly for CSRF and authentication)                                  |

Reviewer: ```suggestion
| GDS_BFF_COOKIE_DOMAIN | string   |                       | The domain to set secure cookies for (particularly for CSRF and authentication).                                  |
```

> | GDS_BFF_SERVE_DOCS    | bool     | false                 | If true, OpenAPI documentation is complied and served alongside the BFF API.                                     |

Reviewer: ```suggestion
| GDS_BFF_SERVE_DOCS    | bool     | false                 | If true, OpenAPI documentation is compiled and served alongside the BFF API.                                     |
```

|-------------------------------|----------|---------|--------------------------------------------------------------------|
| GDS_BFF_USER_CACHE_ENABLED    | bool     | false   | Enable user caching to reduce lookups with the Auth0 API.          |
| GDS_BFF_USER_CACHE_SIZE       | uint     | 16384   | The size in bytes to limit the LRU cache to.                       |
| GDS_BFF_USER_CACHE_EXPIRATION | duration | 8h      | How long to keep records in the cache before forcing a new lookup. |

### Auth0 Config

The BFF uses Auth0 for authentication and authorization and must connect to the Auth0 Management API in order to manage users. The Auth0 client is configured as follows:

| EnvVar                       | Type     | Default | Description                                                                                                                                                                        |
|------------------------------|----------|---------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | GDS_BFF_AUTH0_DOMAIN         | string   |         | The tenant domain provided by the Auth0 application or API (domain only, no scheme or path)                                                                                        |

Reviewer: ```suggestion
| GDS_BFF_AUTH0_DOMAIN         | string   |         | The tenant domain provided by the Auth0 application or API (domain only, no scheme or path).                                                                                        |
```

| GDS_BFF_AUTH0_CLIENT_SECRET  | string   |         | The Client Secret for the management API specified by Auth0.                                                                                                                       |
| GDS_BFF_AUTH0_TESTING        | bool     | false   | If true a mock authenticator is used for testing purposes.                                                                                                                         |

### Network Configuration

The network configuration enables the BFF to connect to the GDS Database, Directory API, and Members API for both the MainNet and TestNet networks. The required configuration and configuration values is the same for both networks but the environment variables are prefixed with `GDS_BFF_TESTNET_` and `GDS_BFF_MAINNET_` respectively.

The configuration without the prefixes is specified below, at the end of each section, we will provide an exhaustive list of environment variables that are required to fully configure the BFF for connecting to both the TestNet and MainNet GDS services.

#### GDS Database Connection

> The GDS Database connection is a store connection that is similar to the BFF Datbase connection described in a following section. Configuring for the TestNet and MainNet database connections is as follows:

Reviewer: ```suggestion
The GDS Database connection is a store connection that is similar to the BFF Database connection described in the following section. Configuration for the TestNet and MainNet database connections is as follows:
```

| EnvVar                   | Type   | Default | Description                                                                          |
|--------------------------|--------|---------|--------------------------------------------------------------------------------------|

> | DATABASE_URL             | string |         | Required, the DSN to connect to the database on (see below for details)              |

Reviewer: ```suggestion
| DATABASE_URL             | string |         | Required, the DSN to connect to the database on (see below for details).              |
```

> | DATABASE_REINDEX_ON_BOOT | bool   | false   | When the server starts, instead of loading indexes from disk, recreate and save them |

Reviewer: ```suggestion
| DATABASE_REINDEX_ON_BOOT | bool   | false   | When the server starts, instead of loading indexes from disk, recreate and save them. |
```

> | DATABASE_INSECURE        | bool   | false   | If set do not connect to the TrtlDB with mTLS authentication                         |

Reviewer: ```suggestion
| DATABASE_INSECURE        | bool   | false   | If set do not connect to the TrtlDB with mTLS authentication.                        |
```

> | DATABASE_CERT_PATH       | string |         | The path to the mTLS client-side certs for database auth                             |

Reviewer: ```suggestion
| DATABASE_CERT_PATH       | string |         | The path to the mTLS client-side certs for database auth.                             |
```

> | DATABASE_POOL_PATH       | string |         | The path to the mTLS public cert pools to accept server connections                  |

Reviewer: ```suggestion
| DATABASE_POOL_PATH       | string |         | The path to the mTLS public cert pools to accept server connections.                  |
```

Note that only a `trtl://` database url should be used for network database connections and that `DATABASE_REINDEX_ON_BOOT` should _always_ be `false` for the BFF.

List of environment variables:

- `GDS_BFF_TESTNET_DATABASE_URL`
- `GDS_BFF_TESTNET_DATABASE_INSECURE`
- `GDS_BFF_TESTNET_DATABASE_CERT_PATH`
- `GDS_BFF_TESTNET_DATABASE_POOL_PATH`
- `GDS_BFF_MAINNET_DATABASE_URL`
- `GDS_BFF_MAINNET_DATABASE_INSECURE`
- `GDS_BFF_MAINNET_DATABASE_CERT_PATH`
- `GDS_BFF_MAINNET_DATABASE_POOL_PATH`

#### GDS Directory API Configuration

The BFF connects to the GDS TRISA Directory API to perform operations like registration submission, contact verification, and verification status lookups. The configuration for the directory API clients is as follows:

| EnvVar             | Type     | Default | Description                                                         |
|--------------------|----------|---------|---------------------------------------------------------------------|

> | DIRECTORY_INSECURE | bool     | false   | If false does not connect to the directory API using TLS            |

Reviewer: ```suggestion
| DIRECTORY_INSECURE | bool     | false   | If false does not connect to the directory API using TLS.            |
```

> | DIRECTORY_ENDPOINT | string   |         | The endpoint (host:port) to connect to the directory API on         |

Reviewer: ```suggestion
| DIRECTORY_ENDPOINT | string   |         | The endpoint (host:port) to connect to the directory API on.         |
```

> | DIRECTORY_TIMEOUT  | duration | 10s     | The connection timeout for directory API request and dial contexts  |

Reviewer: ```suggestion
| DIRECTORY_TIMEOUT  | duration | 10s     | The connection timeout for directory API request and dial contexts.  |
```

List of environment variables:

- `GDS_BFF_TESTNET_DIRECTORY_INSECURE`
- `GDS_BFF_TESTNET_DIRECTORY_ENDPOINT`
- `GDS_BFF_TESTNET_DIRECTORY_TIMEOUT`
- `GDS_BFF_MAINNET_DIRECTORY_INSECURE`
- `GDS_BFF_MAINNET_DIRECTORY_ENDPOINT`
- `GDS_BFF_MAINNET_DIRECTORY_TIMEOUT`

#### GDS Members API Configuration

The BFF connects to the secure GDS Members API to give logged in users access to the complete directory including listing verified members and updating their member record. The configuration for the members API client is as follows:

| EnvVar                 | Type     | Default | Description                                                         |
|------------------------|----------|---------|---------------------------------------------------------------------|

> | MEMBERS_ENDPOINT       | string   |         | The endpoint (host:port) to connect to the members API on           |

Reviewer: ```suggestion
| MEMBERS_ENDPOINT       | string   |         | The endpoint (host:port) to connect to the members API on.           |
```

> | MEMBERS_TIMEOUT        | duration | 10s     | The connection timeout for members API request and dial contexts    |

Reviewer: ```suggestion
| MEMBERS_TIMEOUT        | duration | 10s     | The connection timeout for members API request and dial contexts.    |
```

> | MEMBERS_MTLS_INSECURE  | bool     | false   | If false does not connect to the members API using mTLS             |

Reviewer: ```suggestion
| MEMBERS_MTLS_INSECURE  | bool     | false   | If false does not connect to the members API using mTLS.             |
```

> | MEMBERS_MTLS_CERT_PATH | string   |         | The path to the mTLS client-side certs for members API auth         |

Reviewer: ```suggestion
| MEMBERS_MTLS_CERT_PATH | string   |         | The path to the mTLS client-side certs for members API auth.         |
```

> | MEMBERS_MTLS_POOL_PATH | string   |         | The path to the mTLS public cert pools to accept server connections |

Reviewer: ```suggestion
| MEMBERS_MTLS_POOL_PATH | string   |         | The path to the mTLS public cert pools to accept server connections. |
```

- `GDS_BFF_TESTNET_MEMBERS_MTLS_INSECURE`
- `GDS_BFF_TESTNET_MEMBERS_MTLS_CERT_PATH`
- `GDS_BFF_TESTNET_MEMBERS_MTLS_POOL_PATH`
- `GDS_BFF_MAINNET_MEMBERS_ENDPOINT`
- `GDS_BFF_MAINNET_MEMBERS_TIMEOUT`
- `GDS_BFF_MAINNET_MEMBERS_MTLS_INSECURE`
- `GDS_BFF_MAINNET_MEMBERS_MTLS_CERT_PATH`
- `GDS_BFF_MAINNET_MEMBERS_MTLS_POOL_PATH`

### Database

The BFF makes use of a globally replicated key-value store for persistence of BFF-specific data structures such as organizations and announcements. By default it uses the TrtlDB for this, but for testing or smaller deployments it can use a local LevelDB database instead.

Generally speaking, the BFF uses the same TrtlDB as the GDS MainNet instance, and we've ensured there are no namespace conflicts to prevent this. The BFF primary store can also be independent of the GDS stores if necessary.

The primary database store is configured as follows:

| EnvVar                       | Type   | Default | Description                                                                          |
|------------------------------|--------|---------|--------------------------------------------------------------------------------------|

> | GDS_BFF_DATABASE_URL             | string |         | Required, the DSN to connect to the database on (see below for details)              |

Reviewer: ```suggestion
| GDS_BFF_DATABASE_URL             | string |         | Required, the DSN to connect to the database on (see below for details).              |
```

> | GDS_BFF_DATABASE_REINDEX_ON_BOOT | bool   | false   | When the server starts, instead of loading indexes from disk, recreate and save them |

Reviewer: ```suggestion
| GDS_BFF_DATABASE_REINDEX_ON_BOOT | bool   | false   | When the server starts, instead of loading indexes from disk, recreate and save them. |
```

> | GDS_BFF_DATABASE_INSECURE        | bool   | false   | If set do not connect to the TrtlDB with mTLS authentication                         |

Reviewer: ```suggestion
| GDS_BFF_DATABASE_INSECURE        | bool   | false   | If set do not connect to the TrtlDB with mTLS authentication.                         |
```

> | GDS_BFF_DATABASE_CERT_PATH       | string |         | The path to the mTLS client-side certs for database auth                             |

Reviewer: ```suggestion
| GDS_BFF_DATABASE_CERT_PATH       | string |         | The path to the mTLS client-side certs for database auth.                             |
```

> | GDS_BFF_DATABASE_POOL_PATH       | string |         | The path to the mTLS public cert pools to accept server connections                  |

Reviewer: ```suggestion
| GDS_BFF_DATABASE_POOL_PATH       | string |         | The path to the mTLS public cert pools to accept server connections.                  |
```

When connecting to a LevelDB, the path to the directory on disk where the leveldb should be stored must be passed to the DSN. To specify a relative path, use three slashes: `leveldb:///relpath/to/db`, to specify an absolute path use four: `leveldb:////abspath/to/db`.

### Email and SendGrid

The BFF uses [SendGrid](https://sendgrid.com/) to send email notifications and to enable communication workflows with users and admins. Configure BFF to use SendGrid as follows:

| EnvVar                | Type   | Default                                           | Description                                                              |
|-----------------------|--------|---------------------------------------------------|--------------------------------------------------------------------------|
| GDS_BFF_SERVICE_EMAIL | string | TRISA Directory Service <admin@vaspdirectory.net> | The email address used as the sender for all emails from the BFF system. |
| SENDGRID_API_KEY      | string |                                                   | API Key to authenticate to SendGrid with.                                |

> | GDS_BFF_EMAIL_TESTING | bool   | false                                             | Use email in testing mode rather than send live emails                   |

Reviewer: ```suggestion
| GDS_BFF_EMAIL_TESTING | bool   | false                                             | Use email in testing mode rather than send live emails.                   |
```

> | GDS_BFF_EMAIL_STORAGE | string | ""                                                | Directory to store test emails for "mark one eyeball" review             |

Reviewer: ```suggestion
| GDS_BFF_EMAIL_STORAGE | string | ""                                                | Directory to store test emails for "mark one eyeball" review.             |
```
